fix: don't flush model when we have no machines

adisazhar123 opened the pull request:
<!--
The PR title should match: <type>(optional <scope>): <description>.

Please also ensure all commits in this PR comply with our conventional commits specification:
https://github.com/juju/juju/blob/main/docs/contributor/reference/conventional-commits.md
-->

<!-- Why this change is needed and what it does. -->

When we add a new model without machines, the firewall worker keeps on polling for security groups infinitely. As we don't have machines yet for the model, the security groups aren't created, in turn causing the firewall to keep retrying. This has caused an excessive amount of API calls to Neutron `GET /v2.0/security-groups`.

As a part of this PR, we added a condition to only flush a model (hence fetching security groups) when we have at least one machine, and skipping it otherwise. This check is performed in the firewall worker.

## Checklist

<!-- If an item is not applicable, use `~strikethrough~`. -->

- [ ] Code style: imports ordered, good names, simple structure, etc
- [ ] Comments saying why design decisions were made
- [ ] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

<!--

Describe steps to verify that the change works.

If you're changing any of the facades, you need to ensure that you've tested
a model migration from 3.6 to 4.0 and from 4.0 to 4.0.

The following steps are a good starting point:

 1. Bootstrap a 3.6 controller and deploy a charm.

```sh
$ juju bootstrap lxd src36
$ juju add-model moveme1
$ juju deploy juju-qa-test
```

 2. Bootstrap a 4.0 controller with the changes and migrate the model.

```sh
$ juju bootstrap lxd dst40
$ juju migrate src36:moveme1 dst40
$ juju add-unit juju-qa-test
```

 3. Verify no errors exist in the model logs for the agents. If there are
    errors, this is a bug and should be fixed before merging. The fix can
    either be applied to the 4.0 branch (preferable) or the 3.6 branch, though
    that needs to be discussed with the team.

```sh
$ juju debug-log -m dst40:controller
$ juju debug-log -m dst40:moveme1
```

    4. We also need to test a model migration from 4.0 to 4.0.

```sh
$ juju bootstrap lxd src40
$ juju add-model moveme2
$ juju deploy juju-qa-test
```

```sh
$ juju migrate src40:moveme2 dst40
$ juju add-unit juju-qa-test
```

    5. Verify that there are no errors in the controller or model logs.

```sh
$ juju debug-log -m dst40:controller
$ juju debug-log -m dst40:moveme2
```

-->

## Documentation changes

<!-- How it affects user workflow (CLI or API). -->

## Links

<!-- Link to all relevant specification, documentation, bug, issue or JIRA card. -->

<!-- Replace #19267 with an issue reference or link, otherwise remove this line. -->
**Issue:** Fixes #18649.

<!-- Add a Jira card reference or link, otherwise remove this line. -->
**Jira card:** [JUJU-7455](https://warthogs.atlassian.net/browse/JUJU-7455)

wallyworld commented on internal/worker/firewaller/firewaller.go:
```diff
@@ -858,6 +859,15 @@ func (fw *Firewaller) flushMachine(machined *machineData) error {
 		return fw.flushGlobalPorts(toOpen, toClose)
 	}
+
+	// We may have had a call to flushModel() in the past but did not have any machines yet.
```
// We may have received a notification to ...

wallyworld replied:
Also, do this at the very top, or else it won't always get called

adisazhar123 replied:
Done

wallyworld commented on internal/worker/firewaller/firewaller.go:
```diff
@@ -1095,6 +1105,16 @@ func (fw *Firewaller) flushModel() error {
 		return nil
 	}
+
+	// We may be in a situation where we have added a model but doesn't have machines yet.
+	// To prevent the firewall worker from infinitely polling the Neutron security group API
+	// we skip flushing a model if we have no machines.
```
This is leaking provider specific details. Just say something along the lines of we don't want to create the model specific artefacts until there's at least one machine in the model.

wallyworld commented on internal/worker/firewaller/firewaller.go:
```diff
@@ -1126,6 +1146,9 @@ func (fw *Firewaller) flushModel() error {
 	}
+
+	// Reset the flag once we have flushed the model.
+	fw.needsToFlushModel = false
```
Reset the flag immediately after the machines check. Doing it here doesn't change anything.

adisazhar123 replied:
Done.

wallyworld reviewed (commented):
Looks good. A few small things to get started.

adisazhar123 commented:
@wallyworld  Thanks for the early feedback. Changes have been made.

SimonRichardson commented on internal/worker/firewaller/export_test.go:
```diff
@@ -0,0 +1,9 @@
+package firewaller
```
We've tended to move away from export_test files as it leads to bad testing composition.

This wouldn't be permitted in 4.0, so we shouldn't have it in 3.6 imo.

adisazhar123 replied:
Thanks for the heads up folks. As discussed with @wallyworld, I'll attempt to backport the unit test setup from main. I'll see which commits to cherry pick and what tweaks are required

adisazhar123 commented:
@wallyworld I have backported the unit test setup from main. It should be much easier to reason around the tests now.

Sharing the cherry-picked commits below for visibility. There were tweaks to get it working.

- https://github.com/juju/juju/pull/15804/commits/28bed95c3bb69c8ad285367ee94b525f15d15f84
- https://github.com/juju/juju/pull/15958/commits/8556b9873b8251676f6e52f923b035aae98f0c28
- https://github.com/juju/juju/commit/cf06a8e03f3a62e5aeacee289d273c14544a4826

I would appreciate another round of review, thanks!

wallyworld commented on internal/worker/firewaller/firewaller_test.go:
```diff
@@ -6,131 +6,198 @@ package firewaller_test
+
+	// This is the controller machine.
+
+	if s.createWithMachine {
```
Don't use `createWithMachine` as a struct attribute.
Create a `addMachine()` method (or whatever the name might be) and in that method do the addMachine, startInstance, ensureMocksForMachineCreation calls. 
Have `ensureMocksWithMachine()` and `ensureMocksWithoutMachine()` or whatever naming woks best . The former will call `ensureMocksWithoutMachine` followed by `addMachine` etc

adisazhar123 replied:
Extracted into its own function for mocks without machines

wallyworld commented on internal/worker/firewaller/firewaller_test.go:
```diff
@@ -164,781 +228,968 @@ func (s *firewallerBaseSuite) assertEnvironPorts(c *gc.C, expected firewall.Ingr
-	firewallerBaseSuite
-	watchMachineNotify func(tag names.MachineTag)
-	flushModelNotify   func()
+func (s *firewallerBaseSuite) waitForSkippingModelFlush(c *gc.C) {
```
I don't see why we need this. If the model flush is skipped, then the EXPECT() call sequence will change. ie the test where we expect the model to not be flushed will not include the extra expected calls.

adisazhar123 replied:
As discussed, will remove because we can use the EXPECT() calls.

wallyworld reviewed (commented):
Thanks for the backport of the tests. It's looking good - a few things to fix.

adisazhar123 commented on internal/worker/firewaller/firewaller_test.go:
```diff
@@ -6,131 +6,198 @@ package firewaller_test
+	s.IsolationSuite.SetUpTest(c)
 
-	s.callCtx = context.NewEmptyCloudCallContext()
+	s.modelFlushed = make(chan bool, 5)
+	s.machineFlushed = make(chan string, 5)
+	s.watchingMachine = make(chan names.MachineTag, 5)
+	s.skippedModelFlush = make(chan bool, 5)
```
Done

adisazhar123 commented on internal/worker/firewaller/firewaller.go:
```diff
@@ -109,14 +53,21 @@ type Config struct {
+	// FlushMMachineNotify is called when the Firewaller flushes a machine.
+	// This should only be used for testing
+	FlushMachineNotify func(string)
+	// SkipFlushModelNotify is called when the Firewaller skips flushing it's model.
```
As discussed this function is removed

adisazhar123 commented:
@wallyworld thanks for the feedback. I've pushed some changes:
- check for error in `Unit()`
- removed `SkipFlushModelNotify()` and `NeedsToFlushModel()`
- fixed typos and make types to be consistent
- removed redundant channel instantiation
- added `AddCleanUp()`
- introduced a new function, `addMachineUnitAndEnsureMocks()`, to create machines and methods expectations

Need your help to take a look again please. Tq.

adisazhar123 commented on internal/worker/firewaller/firewaller_test.go:
```diff
@@ -6,131 +6,198 @@ package firewaller_test
+		s.mu.Lock()
+		got := s.instancePorts[machineId]
+		if expected.EqualTo(got) {
 			c.Succeed()
```
Done

wallyworld commented on internal/worker/firewaller/firewaller_test.go:
```diff
@@ -947,183 +1319,161 @@ func (s *InstanceModeSuite) TestStartWithStateOpenPortsBroken(c *gc.C) {
-		CACert:        coretesting.CACert}, offeringModelTag.Id())
-	c.Assert(err, jc.ErrorIsNil)
+	fw := s.newFirewaller(c)
+	defer workertest.CleanKill(c, fw)
```
Don't need defer.

wallyworld commented on internal/worker/firewaller/firewaller_test.go:
```diff
@@ -947,183 +1319,161 @@ func (s *InstanceModeSuite) TestStartWithStateOpenPortsBroken(c *gc.C) {
-		Addrs:         []string{"1.2.3.4:1234"},
-		CACert:        coretesting.CACert}, offeringModelTag.Id())
-	c.Assert(err, jc.ErrorIsNil)
+	fw := s.newFirewaller(c)
```
We want to first add some model ingress rules so that a flush would actually occur right?

eg
```
	s.modelIngressRules = firewall.IngressRules{
		firewall.NewIngressRule(network.MustParsePortRange("22"), firewall.AllNetworksIPV4CIDR, firewall.AllNetworksIPV6CIDR),
		firewall.NewIngressRule(network.MustParsePortRange("17070"), firewall.AllNetworksIPV4CIDR),
	}
```

adisazhar123 replied:
Ah I removed the ingressRules because there is no machine so I feel it doesn't make sense to have them. Instead, I send an empty struct to modelFirewallChanges channel ([ref](https://github.com/juju/juju/blob/4ba9892ff725ad52d59204cb0b4b0766bcf45f1d/internal/worker/firewaller/firewaller_test.go#L213)) to trigger a flush.

wallyworld reviewed (approved):
Thank you

CodingCookieRookie commented on internal/worker/firewaller/firewaller.go:
```diff
@@ -109,14 +53,18 @@ type Config struct {
+	// This should only be used for testing.
 	FlushModelNotify func()
+	// FlushMachineNotify is called when the Firewaller flushes a machine.
+	// This should only be used for testing.
```
Hi Adis, perhaps the line here about using only for test should be removed since we use this func in the below func (fw *Firewaller) flushMachine(machined *machineData)?

CodingCookieRookie reviewed (approved):
LGTM, thanks Adis! Just a comment refactor and friendly reminder to fix the lint :)

adisazhar123 commented:
/merge
